pkg/distributor/validator.go

```diff
@@ -40,26 +44,71 @@ func (v Validator) ValidateEntry(userID string, entry logproto.Entry) error {
 		// but the upstream cortex_validation pkg uses it, so we keep this
 		// for parity.
 		validation.DiscardedSamples.WithLabelValues(validation.LineTooLong, userID).Inc()
+		validation.DiscardedBytes.WithLabelValues(validation.LineTooLong, userID).Add(float64(len(entry.Line)))
 		return httpgrpc.Errorf(
 			http.StatusBadRequest,
-			"max line size (%s) exceeded while adding (%s) size line",
-			flagext.ByteSize(uint64(maxSize)).String(),
-			flagext.ByteSize(uint64(len(entry.Line))).String(),
+			validation.LineTooLongErrorMsg,
+			maxSize,
+			labels,
+			len(entry.Line),
 		)
 	}
 
 	return nil
 }
 
 // Validate labels returns an error if the labels are invalid
-func (v Validator) ValidateLabels(userID string, labels string) error {
-	ls, err := util.ToClientLabels(labels)
+func (v Validator) ValidateLabels(userID string, stream *logproto.Stream) error {
+	ls, err := util.ToClientLabels(stream.Labels)
 	if err != nil {
 		// I wish we didn't return httpgrpc errors here as it seems
 		// an orthogonal concept (we need not use ValidateLabels in this context)
 		// but the upstream cortex_validation pkg uses it, so we keep this
 		// for parity.
 		return httpgrpc.Errorf(http.StatusBadRequest, "error parsing labels: %v", err)
 	}
-	return cortex_validation.ValidateLabels(v, userID, ls)
+
+	numLabelNames := len(ls)
+	if numLabelNames > v.MaxLabelNamesPerSeries(userID) {
+		validation.DiscardedSamples.WithLabelValues(validation.MaxLabelNamesPerSeries, userID).Inc()
+		bytes := 0
+		for _, e := range stream.Entries {
+			bytes += len(e.Line)
+		}
+		validation.DiscardedBytes.WithLabelValues(validation.MaxLabelNamesPerSeries, userID).Add(float64(bytes))
+		return httpgrpc.Errorf(http.StatusBadRequest, validation.MaxLabelNamesPerSeriesErrorMsg, cortex_client.FromLabelAdaptersToMetric(ls).String(), numLabelNames, v.MaxLabelNamesPerSeries(userID))
+	}
+
+	maxLabelNameLength := v.MaxLabelNameLength(userID)
+	maxLabelValueLength := v.MaxLabelValueLength(userID)
+	lastLabelName := ""
+	for _, l := range ls {
+		var errTemplate string
+		var reason string
+		var cause interface{}
```

Comment:
Can this be `string`?

Reply:
yup, this was part of the code copied from cortex and I removed a few of the validations, some of which I think returned numeric types, this can be string now, good catch